astropy/table/sorted_array.py

```diff
@@ -0,0 +1,286 @@
+import numpy as np
+
+class SortedArray(object):
+    '''
+    Implements a sorted array container using
+    a list of numpy arrays.
+
+    Parameters
+    ----------
+    lines : `Table`
+        Sorted columns of the original table as well as
+        a final column consisting of the rows given by argsort()
+    col_dtypes : list or None
+        If not None, col_dtypes should be a list of types for each column
+    data : list or None
+        An internal data object to use instead of creating new data
+    '''
+    def __init__(self, lines, col_dtypes=None, data=None):
+        if data is not None: # sliced reference to data
+            self.length = len(data[0])
+            self._data = data
+            return
+        elif col_dtypes is not None:
```

Comment:
In the function docs the input `lines` is listed as not optional.  If so, then why would you ever need to specify `col_dtypes` explicitly?  That can always be found in the input table of index columns.